src/tdigest/sketch.rs

```diff
@@ -0,0 +1,453 @@
+            self.centroids[num_centroids - 1].mean,
+            w1,
+            self.max,
+            w2,
+        ))
+    }
+
+    /// Process buffered values and merge centroids if needed.
+    pub(super) fn compress(&mut self) {
+        if self.buffer.is_empty() {
+            return;
+        }
+        let mut tmp = Vec::with_capacity(self.buffer.len() + self.centroids.len());
+        for &v in &self.buffer {
+            tmp.push(Centroid { mean: v, weight: 1 });
+        }
+        self.do_merge(tmp, self.buffer.len() as u64)
+    }
+
+    /// Merges the given buffer of centroids into this TDigest.
+    ///
+    /// # Contract
+    ///
+    /// * `buffer` must have at least one centroid.
+    /// * `buffer` is generated from `self.buffer`, and thus:
+    ///     * No `NAN` values are present in `buffer`.
+    ///     * We should clear `self.buffer` after merging.
+    pub(super) fn do_merge(&mut self, mut buffer: Vec<Centroid>, weight: u64) {
+        buffer.extend(std::mem::take(&mut self.centroids));
+        buffer.sort_by(centroid_cmp);
+        if self.reverse_merge {
+            buffer.reverse();
+        }
+        self.centroids_weight += weight;
+
+        let mut num_centroids = 0;
+        let len = buffer.len();
+        self.centroids.push(buffer[0]);
+        num_centroids += 1;
+        let mut current = 1;
+        let mut weight_so_far = 0.;
+        while current < len {
+            let c = buffer[current];
+            let proposed_weight = (self.centroids[num_centroids - 1].weight + c.weight) as f64;
+            let mut add_this = false;
+            if (current != 1) && (current != (len - 1)) {
+                let centroids_weight = self.centroids_weight as f64;
+                let q0 = weight_so_far / centroids_weight;
+                let q2 = (weight_so_far + proposed_weight) / centroids_weight;
+                let normalizer = scale_function::normalizer((2 * self.k) as f64, centroids_weight);
+                add_this = proposed_weight
+                    <= (centroids_weight
+                        * scale_function::max(q0, normalizer)
+                            .min(scale_function::max(q2, normalizer)));
+            }
+            if add_this {
+                // merge into existing centroid
+                self.centroids[num_centroids - 1].add(c);
+            } else {
+                // copy to a new centroid
+                weight_so_far += self.centroids[num_centroids - 1].weight as f64;
+                self.centroids.push(c);
+                num_centroids += 1;
+            }
+            current += 1;
+        }
+
+        if self.reverse_merge {
+            self.centroids.reverse();
+        }
+        self.min = self.min.min(self.centroids[0].mean);
+        self.max = self.max.max(self.centroids[num_centroids - 1].mean);
+        self.reverse_merge = !self.reverse_merge;
+        self.buffer.clear();
+    }
+}
+
+fn centroid_cmp(a: &Centroid, b: &Centroid) -> std::cmp::Ordering {
```

Comment:
Can we implement this via Ord / PartialOrd?

Reply:
I moved all the cmp logics together in 9497d24.

Although there is `f64::total_cmp`, it takes `+0.0 > -0.0` which is different from [`OrderedFloat`](https://docs.rs/ordered-float/latest/ordered_float/struct.OrderedFloat.html) which is more widely used and give an "intuitive" total order.

Perhaps we can use OrderedFloat in TDigest. Let me think of it ...